keps/sig-node/5224-node-resource-discovery/README.md

```diff
@@ -0,0 +1,1211 @@
+- Change existing behavior on node capacity changes (covered e.g. [KEP-3953][kep-3953])
+- Ephemeral storage
+
+## Proposal
+
+This proposal suggests the CRI runtime as the source of truth for resource
+capacity. It has visibility and control over all running containers and the
+resources in the host. Also, the runtime could be used to serve other clients
+(e.g. docker) in addition to Kubernetes. CRI runtimes (containerd and cri-o)
+have a plugin mechanism, NRI, which can be extended to provide additional
+service to report resources, and external resource discovery can then be
+deployed as an [NRI][nri] plugin.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a system administrator I want to partition my node between system,
+non-Kubernetes-managed workloads and Kubernetes-managed workloads.
+
+#### Story 2
+
+As a cluster operator I want to hide the true hardware details, e.g. actual
+number of logical CPU cores from the Kubernetes. I want to do this to optimize
+application performance by taking advantage of the detailed knowledge of how
+the hardware works behind the scenes.
+
+
+#### Story 3
+
+As a system administrator, I want my cluster to be aware of the compute
+available throughout its lifespan and improve scheduling based on the node’s
+compute capabilities.
+
+#### Story 4
+
+As a system administrator, I want the cluster to gain insights of the host’s
+architecture specific capabilities and leverage the same for improved
+performance.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Reported capacity does not fit currently running pods. This is more of a
+concern in of [KEP-3953][kep-3953]. In practice, the kubelet/Kubernetes cannot
+prevent resources changing (e.g. CPUs getting offlined).
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### CRI API
+
+A new streaming RPC is added to get the node resources.
+
+The proposed API for exposing resources takes inspiration from
+[NodeResourceTopology API][noderesourcetopology-api]. It allows
+representing the resources in a tree structure representing the hardware
+topology of the system.
+
+```proto
+import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+// GetDynamicRuntimeConfig is a streaming interface for receiving dynamically
+// changing runtime and node configuration
+rpc GetDynamicRuntimeConfig(DynamicRuntimeConfigRequest) returns (stream DynamicRuntimeConfigResponse) {}
```

Comment:
hmm this doesn't really seem like a runtime config..

Reply:
This name came into existence with a background thought of creating a more generic "channel" for the runtime to inform kubelet about changes (without kubelet polling). But I agree, the name is bad and the background idea probably too.

Related, @ffromani suggested having a completely separate service (e.g. `DiscoveryService`, in addition to `RuntimeService` and `ImageService`) for handling this.

Thoughts?

Reply:
yeah or maybe `ResourceService` or `NodeService` or something?

Reply:
Check, I will change this in the next update. I don't have strong opinions on this. One detail is that with a separate service there is the possibility to connect to a 3rd-party NodeService/ResourceService agent. That may be a good thing for some special users(?)